File: baekjoon/python/friends_network_4195.py
# Title: 친구 네트워크
# Link: https://www.acmicpc.net/problem/4195

import sys
from collections import defaultdict

read_single_int = lambda: int(sys.stdin.readline().strip())
read_list_words = lambda: sys.stdin.readline().strip().split(' ')

class UnionFind:
    def __init__(self, n: int):
        self.sizes = [1] * n
        self.rank = [1] * n
        self.parent = [i for i in range(n)]

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int) -> None:
        xset = self.find(x)
        yset = self.find(y)

        if xset == yset:
            return

        if self.rank[xset] < self.rank[yset]:
            self.parent[xset] = yset
            self.sizes[yset] += self.sizes[xset]
        elif self.rank[xset] > self.rank[yset]:
            self.parent[yset] = xset
            self.sizes[xset] += self.sizes[yset]
        else:
            self.parent[yset] = xset
            self.rank[xset] = self.rank[yset] + 1
            self.sizes[xset] += self.sizes[yset]
    
    def get_size(self, x: int) -> int:
        return self.sizes[self.find(x)]


def main():
    t = read_single_int()
    for _ in range(t):
        number = [0]
        def set_num() -> int:
            number[0] += 1
            return number[0]
        name_to_number = defaultdict(set_num)
        f = read_single_int()
        uf = UnionFind(2*f+1)
        for _ in range(f):
            a, b = read_list_words()
            uf.union(name_to_number[a], name_to_number[b])
            print(f'{uf.get_size(name_to_number[a])}')

if __name__ == '__main__':
    main()

File: baekjoon/rust/friends_network_4195.rs
use std::io::{stdin, stdout, BufRead, BufReader, BufWriter, Write};
use std::collections::HashMap;

struct UnionFind {
    sizes: Vec<usize>,
    rank: Vec<usize>,
    parent: Vec<usize>
}

impl UnionFind {
    fn new(n: usize) -> UnionFind {
        UnionFind {
            sizes : vec![1; n],
            rank : vec![1; n],
            parent : (0..n).collect(),
        }
    }

    fn find(&mut self, x: usize) -> usize {
        if self.parent[x] != x {
            self.parent[x] = self.find(self.parent[x]);
        }
        self.parent[x]
    }

    fn union(&mut self, x: usize, y:usize) {
        let xset = self.find(x);
        let yset = self.find(y);

        if xset == yset {
            return;
        }

        if self.rank[xset] < self.rank[yset] {
            self.parent[xset] = yset;
            self.sizes[yset] += self.sizes[xset];
        } else if self.rank[xset] > self.rank[yset] {
            self.parent[yset] = xset;
            self.sizes[xset] += self.sizes[yset];
        } else {
            self.parent[yset] = xset;
            self.rank[xset] = self.rank[yset] + 1;
            self.sizes[xset] += self.sizes[yset];
        }
    }

    fn get_size(&mut self, x: usize) -> usize {
        let i = self.find(x);
        self.sizes[i]
    }
}


fn main() {
    let mut buf = BufReader::new(stdin());
    let mut out = BufWriter::new(stdout());
    let t: u128;
    let mut line = String::new();
    
    buf.read_line(&mut line).unwrap();
    t = line.trim().parse().unwrap();

    for _ in 0..t {
        let mut name_to_int = HashMap::<String, usize>::new();
        let f: usize;
        let mut line = String::new();
        buf.read_line(&mut line).unwrap();
        f = line.trim().parse().unwrap();

        let mut disjoint = UnionFind::new(2*f+1);

        for _ in 0..f {
            let mut line = String::new();
            buf.read_line(&mut line).unwrap();
            let friends = line
                            .split_whitespace()
                            .map(|x| String::from(x))
                            .collect::<Vec<_>>();
            
            match name_to_int.get(&friends[0]) {
                Some(_) => (),
                None => {
                   name_to_int.insert(friends[0].to_string(), name_to_int.len());
                }
            }
 
            match name_to_int.get(&friends[1]) {
                Some(_) => (),
                None => {
                    name_to_int.insert(friends[1].to_string(), name_to_int.len());
                }
            }

            disjoint.union(*name_to_int.get(&friends[0]).unwrap(), *name_to_int.get(&friends[1]).unwrap());
            writeln!(out, "{}", disjoint.get_size(*name_to_int.get(&friends[0]).unwrap())).unwrap();
        }
    }
    out.flush().unwrap();
}